docs/design_doc.md

```diff
@@ -0,0 +1,75 @@
+# Design document
+
+## Goals
+
+* Redesign the Github [mirror](#mirror) to reduce the number of Github API requests required to build up the mirror cache with every deployment. 
+
+* Cache rebuilding time after redeployment should be constant and not grow linearly in the number of [mirror replicas](#mirror-replica). 
+
+* Minimize latency so that the average mirror [request time](#request-time) does not increase significantly.
+
+## Current solution
+
+Each replica of the github mirror server maintains its own [in-memory cache](#in-memory-cache). The cache is a data structure (Python dictionary) with constant-time lookup. The cache may grow unbounded if the pod does not fail or get redeployed.
+
+The current design incurs a significant overhead during redeployment as each replica has to build and maintain its own copy of the cache. This design does not scale well if the number of replicas is increased, as the [hourly quota](https://developer.github.com/v3/#rate-limiting) for Github API requests will quickly be exhausted, causing a hold up for all [clients](#client) depending on this service.
+
+## Proposed solution
+
+* Replace each pod's in-memory cache with a single shared cache for all pods. 
+
+    * The [cache server](#cache-server) should be deployed within the same Open Shift cluster as the Github mirror to reduce the latency associated with requests to the cache.
+
+    * A shared cache should reduce the number of calls to the Github API since each pod will not maintain its own copy of the cache.
+
+    * When the mirror is redeployed, the cache will have to be rebuilt once only, instead of once for *each* replica.
```

Comment:
good question. well my thinking was that Redis is also deployed on OpenShift, and if the Redis pod is restarted then we will lose the contents of the cache since Redis stores the data in-memory. we could persist the data by writing to disk, but the disk will be lost if the pod is restarted (specifically I'm thinking if one pod fails, another is spun up immediately and we start over). does that answer the question?

Reply:
Not really :) The are a few things here:

* If we separate the mirror pods and the cache pods, if you restart the mirror the cache is there, so there should be no need to build it again
* If we choose to persist Redis, it will be persisted in a k8s PV (an EBS volume). If the Redis pod is deleted, a new pod will be created that will get attached to the existing PV as Redis will be deployed as a stateful set in k8s

Reply:
There's the ephemeral disk that the pod uses when it's running and there are the storage volumes mounted into the pods at runtime that are kept over pod restarts.
We can limit the scope os this proposal to not persisting data, since it would involve setting up the storage backend (EBS being a great option), the K8S PersistentVolumes objects, the corresponding VolumeClaims and the volumes in the Deployments.
I'd keep it as a future improvement and work on it if there's time left.

Reply:
All that should come automatically once you deploy Redis as a StatefulSet, which is what you have in projects to deploy it in HA mode such as https://github.com/DandyDeveloper/charts/tree/master/charts/redis-ha

I do agree that Redis proper setup is involved, though

Reply:
I considered using AWS Elasticache Redis but that would have cost more, and one goal was to do this without additional cost or vendor lock-in, if possible. I think that periodically snapshotting and saving the snapshots to S3 can still allow us to preserve most of the cache even if that means some writes are lost. The persistence is more complicated and adds comparatively little value (but perhaps I am missing a piece of the bigger picture here). Starting with a simpler design and then iterating on it may be a better strategy to ensure that I can deliver a minimum viable product. What do you think?

Reply:
I'm not worried about vendor lock in. What AWS Elasticache Redis offers is vanilla Redis. It just saves you the hassle of maintaining it :)

I like the periodic snapshotting approach.